components/scream/src/share/util/scream_kokkos_utils.hpp

```diff
@@ -117,61 +121,158 @@ struct ExeSpaceUtils<Kokkos::Cuda> {
   KOKKOS_INLINE_FUNCTION
-  int get_workspace_idx(const MemberType& /* team_member */) const
-  {
+  int get_workspace_idx(const MemberType& /*team_member*/) const
+  { return 0; }
+
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION
+  void release_workspace_idx(const MemberType& /*team_member*/, int /*ws_idx*/) const
+  { }
+};
+
+template <typename ExeSpace = Kokkos::DefaultExecutionSpace>
+class TeamUtils : public _TeamUtilsCommonBase<ExeSpace>
+{
+ public:
+  template <typename TeamPolicy>
+  TeamUtils(const TeamPolicy& policy, const Real& = 1.0) :
+    _TeamUtilsCommonBase<ExeSpace>(policy)
+  { }
+};
+
+/*
+ * Specialization for OpenMP execution space
+ */
 #ifdef KOKKOS_ENABLE_OPENMP
-    return omp_get_thread_num() / _team_size;
-#else
-    return 0;
-#endif
-  }
+template <>
+class TeamUtils<Kokkos::OpenMP> : public _TeamUtilsCommonBase<Kokkos::OpenMP>
+{
+ public:
+  template <typename TeamPolicy>
+  TeamUtils(const TeamPolicy& policy, const Real& = 1.0) :
+    _TeamUtilsCommonBase<Kokkos::OpenMP>(policy)
+  { }
+
+  template <typename MemberType>
+  KOKKOS_INLINE_FUNCTION
+  int get_workspace_idx(const MemberType& /*team_member*/) const
+  { return omp_get_thread_num() / _team_size; }
 };
+#endif
 
 /*
- * Specialization for Cuda execution space. On GPU, we expect all teams to run
- * at once.
+ * Specialization for Cuda execution space.
  */
 #ifdef KOKKOS_ENABLE_CUDA
 template <>
-class TeamUtils<Kokkos::Cuda>
+class TeamUtils<Kokkos::Cuda> : public _TeamUtilsCommonBase<Kokkos::Cuda>
 {
-  int _num_teams;
+  using Device = Kokkos::Device<Kokkos::Cuda, typename Kokkos::Cuda::memory_space>;
+  using flag_type = int; // this appears to be the smallest type that correctly handles atomic operations
+  using view_1d = typename KokkosTypes<Device>::view_1d<flag_type>;
+  using RandomGenerator = Kokkos::Random_XorShift64_Pool<Kokkos::Cuda>;
+  using rnd_type = typename RandomGenerator::generator_type;
 
-public:
+  int             _num_ws_slots;    // how many workspace slots (potentially more than the num of concurrent teams due to overprovision factor)
+  bool            _need_ws_sharing; // true if there are more teams in the policy than ws slots
+  view_1d         _open_ws_slots;    // indexed by ws-idx, true if in current use, else false
+  RandomGenerator _rand_pool;
+
+ public:
   template <typename TeamPolicy>
-  TeamUtils(const TeamPolicy& policy) { _num_teams = policy.league_size(); }
+  TeamUtils(const TeamPolicy& policy, const Real& overprov_factor = 1.0) :
+    _TeamUtilsCommonBase<Kokkos::Cuda>(policy),
+    _num_ws_slots(_league_size > _num_teams
+                  ? (overprov_factor * _num_teams > _league_size ? _league_size : overprov_factor * _num_teams)
+                  : _num_teams),
+    _need_ws_sharing(_league_size > _num_ws_slots),
+    _open_ws_slots("open_ws_slots", _need_ws_sharing ? _num_ws_slots : 0),
+    _rand_pool(std::chrono::high_resolution_clock::now().time_since_epoch().count())
```

Comment:
Is there a way to avoid initializing _rand_pool unless _need_ws_sharing, but without making _rand_pool a pointer?

Reply:
@ambrad, it looks like the default constructor for Random_XorShift64_Pool does basically nothing:
```
    KOKKOS_INLINE_FUNCTION
    Random_XorShift64_Pool() {
      num_states_ = 0;
    }
```

The class also supports assignment. So, we can default construct it and then assign it to a real pool if needed.

Reply:
Great.